perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ExtSourcesManagerBlImpl.java

```diff
@@ -62,67 +74,111 @@ public ExtSourcesManagerBlImpl(ExtSourcesManagerImplApi extSourcesManagerImpl) {
 	@Override
 	public void initialize(PerunSession sess) {
 		if (!this.initialized.compareAndSet(false, true)) return;
+		ExtSourceGoogle.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceVOOT.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceUnity.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceCSV.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceISMU.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceLdap.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceSql.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourcePerun.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceREMS.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceTCS.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceSqlComplex.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceXML.setPerunBlImpl((PerunBlImpl) perunBl);
 		this.extSourcesManagerImpl.initialize(sess);
+		cacheExtSourcesInMemory(sess);
+	}
+
+	private Map<Integer,ExtSource> extSourcesByIdMap = new HashMap<>();
+	private Map<String,ExtSource> extSourcesByNameMap = new HashMap<>();
+	
+	private void cacheExtSourcesInMemory(PerunSession sess) {
+		Map<Integer,ExtSource> extSourcesByIdMap = new HashMap<>();
+		Map<String,ExtSource> extSourcesByNameMap = new HashMap<>();
+		// read all ExtSources from database
+		List<ExtSource> extSources = getExtSourcesManagerImpl().getExtSources(sess);
+		// cache them in hashmaps
+		for(ExtSource extSource : extSources) {
+			extSourcesByIdMap.put(extSource.getId(), extSource);
+			extSourcesByNameMap.put(extSource.getName(), extSource);
+		}
+		this.extSourcesByIdMap = extSourcesByIdMap;
+		this.extSourcesByNameMap = extSourcesByNameMap;
```

Comment:
We should consider of locking these maps while changing to prevent race conditions. What do you think? At least there is a small window when extSourceByIdMap could be different from extSourceByNameMap.

Reply:
I think there cannot be a race condition. Assigning is atomic, so any code sees either the state before reloading, or after reloading.